### PR TITLE
Change event queue process

### LINKED ISSUE #19701: Release 4.7.0 - Alpha 1 - Demo use cases
### Demo use cases information
|                                  |                                                                 |
|--------------------------------- |--------------------------------------------                     |
| **Main release candidate issue** |  #19667                                    |
| **Version**                      | 4.7.0                                                   |
| **Release candidate**            | Alpha 1                                                     |
| **Tag**                          | https://github.com/wazuh/wazuh/tree/v4.7.0-alpha1          |
| **Previous Demo use cases**      | #19486                                                   |

---

## Checks
Status | Use case               | Issues |
:--:   | :--                     | :--:    |
:green_circle:     | Audit                  |        |
:green_circle:     | AWS Wodle              |        | 
:green_circle:     | Brute force            |        |
:green_circle:     | Docker                 |        |
:red_circle:     | Emotet                 |  https://github.com/wazuh/internal-devel-requests/issues/382 | 
:green_circle:     | FIM                    |        |
:green_circle:     | IP Reputation          |        | 
:green_circle:     | Netcat                 |        | 
:green_circle:     | Osquery                |        | 
:green_circle:     | Shellshock             |        |  
:green_circle:     | SQL Injection          |        | 
:green_circle:     | Slack                  |        | 
:green_circle:     | Suricata               |        |   
:green_circle:     | Trojan                 |        |
:green_circle:     | Virustotal             |        | 
:green_circle:     | Vulnerability Detector |        |
:green_circle:     | Yara                   |        | 
:red_circle:     | Windows Defender       | https://github.com/wazuh/internal-devel-requests/issues/381 | 

---

Status legend:
:black_circle: - Pending/In progress
:white_circle: - Skipped
:red_circle: - Rejected
:yellow_circle: - Ready to review
:green_circle: - Approved

---

## Conclusion
The deployment succeeded but some of the use case tests failed. An issue per failed test was created.

## Auditor's validation

In order to close and proceed with the release or the next candidate version, the following auditors must give the green light to this RC.

- [ ] @davidjiglesias 
- [x] @teddytpc1

### LINKED ISSUE #22837: Missing vulnerabilities if manager is restarted during vulnerability scan
|Wazuh version|Component|Install type|Install method|Platform|
|---|---|---|---|---|
| 4.8.0-beta5 | Vulnerability Scan | ManageAgent | Packages | Ubuntu 22 |

### Description

During a vulnerability scan, if the connection with the agent is lost or if the manager undergoes a force stop, certain vulnerabilities may go unreported and become irrecoverably lost.

### Steps to reproduce

- Deploy a manager and an agent
- Wait until feeds are downloaded in the manager
- Just after the agent registration and the vulnerability scanner start for the new agent, restart the manager several times
- Wait until VD finishes to scan the agent
```
2024/04/08 14:19:34 wazuh-modulesd:vulnerability-scanner[89725] osScanner.hpp:336 at handleRequest(): DEBUG: Vulnerability scan for OS 'Ubuntu' on Agent '013' has completed.
2024/04/08 14:19:34 wazuh-modulesd:vulnerability-scanner[89725] scanOrchestrator.hpp:167 at run(): DEBUG: Event type: 3 processed
```
- Check that some vulnerabilities are missing 

### Evidences

> [!NOTE]
> ubuntu_agent_8 and ubuntu_agent_2 have the same packages. `ubuntu_agent_8` was the agent used for reproducing the issue

- Ubuntu agent 2 and Ubuntu agent 8 vulnerabilities: [VulnsForceStop.zip](https://github.com/wazuh/wazuh/files/14906783/VulnsForceStop.zip)
- Manager Logs: [manager_ossec.zip](https://github.com/wazuh/wazuh/files/14906802/manager_ossec.zip)
- Agent Logs: [agent_ossec.zip](https://github.com/wazuh/wazuh/files/14906800/agent_ossec.zip)

### PR DESCRIPTION
**Pull request by Dwordcito**
|Related issue|
|---|
|Close #22837 |

## Description
This pr changes from multiple event queue to process messages to a single queue to have a single thread to process messages.
This aims to have sequential processing of messages.

### CODE REVIEW

File: src/wazuh_modules/vulnerability_scanner/src/scanOrchestrator/scanOrchestrator.hpp [hunk 1]
```diff
@@ -128,46 +143,139 @@ class TScanOrchestrator final : public TOSPrimitives
...
+            [this](rocksdb::PinnableSlice& element)
+            {
+                try
+                {
+                    processEvent(element, true);
+                }
+                catch (const WdbDataException& e)
+                {
+                    m_eventDelayedDispatcher->postpone(e.agentId(), std::chrono::seconds(DELAYED_POSTPONE_SECONDS));
+                    throw std::runtime_error(e.what());
```
Comment (sebasfalcone) [inline src/wazuh_modules/vulnerability_scanner/src/scanOrchestrator/scanOrchestrator.hpp:167]: NitpicK: We could add the agentId to improve the logging down the line
```suggestion
                    throw std::runtime_error(std::string(e.what()) + " (Agent: " + e.agentId() + ")");
```

**Review by sebasfalcone — COMMENTED** [review]
Changes are looking good! A lot of life quality of life improvements 

I'am still testing this, but I leave you some feedback on the code

**Review by sebasfalcone — COMMENTED** [review]
## Testing
If we restart the manager while scanning, this error appears and the process must be killed:
```
2024/04/15 11:08:23 wazuh-db: ERROR: DB((null)) Can't begin transaction.
2024/04/15 11:08:23 wazuh-db: WARNING: DB(queue/db/(null).db) not found. This behavior is unexpected, the database will be recreated.
2024/04/15 11:08:23 wazuh-analysisd: ERROR: dbsync: Bad response from database: Cannot save Syscheck
2024/04/15 11:08:23 wazuh-db: ERROR: DB((null)) Can't begin transaction.
2024/04/15 11:08:23 wazuh-db: WARNING: DB(queue/db/(null).db) not found. This behavior is unexpected, the database will be recreated.
2024/04/15 11:08:23 wazuh-analysisd: ERROR: dbsync: Bad response from database: Cannot save Syscheck
```

The same logs appear from time to time when a syscollector event is received:
```
2024/04/15 11:17:05 wazuh-modulesd:syscollector[134575] logging_helper.c:40 at taggedLogFunction(): DEBUG: Sync sent: {"component":"syscollector_packages","data":{"attributes":{"architecture":"amd64","checksum":"36ae234018ac4f6e7971195df49f6ee0d5495b24","description":"Foreign Function Interface library (development files)","format":"deb","groups":"libdevel","install_time":" ","item_id":"46336a8173721aec176faf860cedfd16a96bb917","location":" ","multiarch":"same","name":"libffi-dev","priority":"optional","scan_time":"2024/04/15 09:17:05","size":328,"source":"libffi","vendor":"Ubuntu Developers <ubuntu-devel-discuss@lists.ubuntu.com>","version":"3.4.2-4"},"index":"46336a8173721aec176faf860cedfd16a96bb917","timestamp":""},"type":"state"}
2024/04/15 11:17:06 wazuh-db: ERROR: DB((null)) Can't begin transaction.
2024/04/15 11:17:06 wazuh-db: WARNING: DB(queue/db/(null).db) not found. This behavior is unexpected, the database will be recreated.
2024/04/15 11:17:06 wazuh-analysisd: ERROR: dbsync: Bad response from database: Cannot save Syscheck
2024/04/15 11:17:06 wazuh-db: ERROR: DB((null)) Can't begin transaction.
2024/04/15 11:17:06 wazuh-db: WARNING: DB(queue/db/(null).db) not found. This behavior is unexpected, the database will be recreated.
2024/04/15 11:17:06 wazuh-analysisd: ERROR: dbsync: Bad response from database: Cannot save Syscheck
2024/04/15 11:17:06 wazuh-db: ERROR: DB((null)) Can't begin transaction.
2024/04/15 11:17:06 wazuh-db: WARNING: DB(queue/db/(null).db) not found. This behavior is unexpected, the database will be recreated.
2024/04/15 11:17:06 wazuh-analysisd: ERROR: dbsync: Bad response from database: Cannot save Syscheck
2024/04/15 11:17:06 wazuh-modulesd:syscollector[134575] logging_helper.c:40 at taggedLogFunction(): DEBUG: Sync sent: {"component":"syscollector_packages","data":{"attributes":{"architecture":"amd64","checksum":"bcae9789a25252498690b334a6a209264d623e9f","description":"multimedia player and streamer (extra access plugins)","format":"deb","groups":"video","install_time":" ","item_id":"4651e238bf5498e77c218f44b18fe0a9fa5d4b66","location":" ","multiarch":"same","name":"vlc-plugin-access-extra","priority":"optional","scan_time":"2024/04/15 09:17:05","size":223,"source":"vlc","vendor":"Ubuntu Developers <ubuntu-devel-discuss@lists.ubuntu.com>","version":"3.0.16-1build7"},"index":"4651e238bf5498e77c218f44b18fe0a9fa5d4b66","timestamp":""},"type":"state"}
2024/04/15 11:17:06 wazuh-modulesd:syscollector[134575] logging_helper.c:40 at taggedLogFunction(): DEBUG: Sync sent: {"component":"syscollector_packages","data":{"attributes":{"architecture":" ","checksum":"a31ef183e3dcab80d327a901e66d2571cd6b33ea","description":" ","format":"pypi","groups":" ","install_time":" ","item_id":"465dc313b5af8e21441bdaa08ee011607cf28f08","location":"/home/sebas/.local/lib/python3.10/site-packages/pyspnego-0.10.2.dist-info/METADATA","name":"pyspnego","priority":" ","scan_time":"2024/04/15 09:17:06","size":0,"source":" ","vendor":" ","version":"0.10.2"},"index":"465dc313b5af8e21441bdaa08ee011607cf28f08","timestamp":""},"type":"state"}
2024/04/15 11:17:06 wazuh-modulesd:syscollector[134575] logging_helper.c:40 at taggedLogFunction(): DEBUG: Sync sent: {"component":"syscollector_packages","data":{"attributes":{"architecture":" ","checksum":"4afa49081c3424f47ed395a21012b5d4927a4cdc","description":" ","format":"pypi","groups":" ","install_time":" ","item_id":"46615767238e9b3dc576ee4a827ce3322981af72","location":"/usr/lib/python3/dist-packages/webencodings-0.5.1.egg-info/PKG-INFO","name":"webencodings","priority":" ","scan_time":"2024/04/15 09:17:06","size":0,"source":" ","vendor":" ","version":"0.5.1"},"index":"46615767238e9b3dc576ee4a827ce3322981af72","timestamp":""},"type":"state"}
2024/04/15 11:17:06 wazuh-modulesd:syscollector[134575] logging_helper.c:40 at taggedLogFunction(): DEBUG: Sync sent: {"component":"syscollector_packages","data":{"attributes":{"architecture":"amd64","checksum":"b075e88277abd1aef92c7c67333cbf8785250399","description":"Asynchronous event notification library (core)","format":"deb","groups":"libs","install_time":" ","item_id":"467ecaf0d88545a68e338f7f2a75bd7b7acb52d8","location":" ","multiarch":"same","name":"libevent-core-2.1-7","priority":"optional","scan_time":"2024/04/15 09:17:06","size":256,"source":"libevent","vendor":"Balint Reczey <rbalint@ubuntu.com>","version":"2.1.12-stable-1build3"},"index":"467ecaf0d88545a68e338f7f2a75bd7b7acb52d8","timestamp":""},"type":"state"}
2024/04/15 11:17:06 wazuh-modulesd:syscollector[134575] logging_helper.c:40 at taggedLogFunction(): DEBUG: Sync sent: {"component":"syscollector_packages","data":{"attributes":{"architecture":" ","checksum":"b8696ac43bd53a3522f3263a13e99314d600510a","description":" ","format":"pypi","groups":" ","install_time":" ","item_id":"4687cb839aca0f56a725f900a61b6164f9ee4cac","location":"/usr/lib/python3/dist-packages/urllib3-1.26.5.egg-info/PKG-INFO","name":"urllib3","priority":" ","scan_time":"2024/04/15 09:17:06","size":0,"source":" ","vendor":" ","version":"1.26.5"},"index":"4687cb839aca0f56a725f900a61b6164f9ee4cac","timestamp":""},"type":"state"}
2024/04/15 11:17:06 wazuh-db: WARNING: DB(queue/db/(null).db) not found. This behavior is unexpected, the database will be recreated.
2024/04/15 11:17:06 wazuh-analysisd: ERROR: dbsync: Bad response from database: Cannot save Syscollector
```

File: src/shared_modules/utils/rocksDBOptions.hpp [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+        /**
+         * @brief Builds the column family options for the RocksDB instance.
+         * @return rocksdb::ColumnFamilyOptions Column family options.
+         */
+        static rocksdb::ColumnFamilyOptions buildColumnFamilyOptions(const std::shared_ptr<rocksdb::Cache>& readCache)
+        {
+            rocksdb::ColumnFamilyOptions columnFamilyOptions;
+            // Amount of data to build up in memory (backed by an unsorted log
+            // on disk) before converting to a sorted on-disk file.
+            columnFamilyOptions.write_buffer_size = 64 * 1024 * 1024;
```
Comment (GabrielEValenzuela) [inline src/shared_modules/utils/rocksDBOptions.hpp:51]: We can avoid the calculation in runtime

```cpp
constexpr uint32_t BUFFER_SIZE = 64 * 1024 * 1024
```


```suggestion
            columnFamilyOptions.write_buffer_size = BUFFER_SIZE;
```

File: src/shared_modules/utils/rocksDBOptions.hpp [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+            // The maximum number of write buffers that are built up in memory.
+            columnFamilyOptions.max_write_buffer_number = 2;
+            // The maximum number of levels of compaction to allow.
+            columnFamilyOptions.num_levels = 4;
```
Comment (GabrielEValenzuela) [inline src/shared_modules/utils/rocksDBOptions.hpp:55]:
```cpp
constexpr uint32_t MAX_WRITERS = 2;
constexpr uint32_t LEVELS = 4;
```

```suggestion
            columnFamilyOptions.max_write_buffer_number = MAX_WRITERS;
            // The maximum number of levels of compaction to allow.
            columnFamilyOptions.num_levels = LEVELS;
```

File: src/shared_modules/utils/rocksDBOptions.hpp [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+            columnFamilyOptions.table_factory.reset(rocksdb::NewBlockBasedTableFactory(buildTableOptions(readCache)));
+
+            return columnFamilyOptions;
+        }
+
+        /**
+         * @brief Builds the DB options for the RocksDB instance.
+         * @return rocksdb::Options DB options.
+         */
+        static rocksdb::Options buildDBOptions(const std::shared_ptr<rocksdb::WriteBufferManager>& writeManager,
```
Comment (GabrielEValenzuela) [inline src/shared_modules/utils/rocksDBOptions.hpp:66]: Ditto

File: src/shared_modules/utils/rocksDBQueueCF.hpp [hunk 5]
```diff
@@ -0,0 +1,350 @@
...
+    }
+
+public:
+    explicit RocksDBQueueCF(const std::string& path)
+    {
+        // RocksDB initialization.
+        // Read cache is used to cache the data read from the disk.
+        m_readCache = rocksdb::NewLRUCache(16 * 1024 * 1024);
+        // Write buffer manager is used to manage the memory used for writing data to the disk.
+        m_writeManager = std::make_shared<rocksdb::WriteBufferManager>(64 * 1024 * 1024);
```
Comment (GabrielEValenzuela) [inline src/shared_modules/utils/rocksDBQueueCF.hpp:134]: Ditto

**Review by GabrielEValenzuela — COMMENTED** [review]
Great job Octa! A few comments!

**Comment by Dwordcito** [comment]
> ## Testing
> If we restart the manager while scanning, this error appears and the process must be killed:
> 
> ```
> 2024/04/15 11:08:23 wazuh-db: ERROR: DB((null)) Can't begin transaction.
> 2024/04/15 11:08:23 wazuh-db: WARNING: DB(queue/db/(null).db) not found. This behavior is unexpected, the database will be recreated.
> 2024/04/15 11:08:23 wazuh-analysisd: ERROR: dbsync: Bad response from database: Cannot save Syscheck
> 2024/04/15 11:08:23 wazuh-db: ERROR: DB((null)) Can't begin transaction.
> 2024/04/15 11:08:23 wazuh-db: WARNING: DB(queue/db/(null).db) not found. This behavior is unexpected, the database will be recreated.
> 2024/04/15 11:08:23 wazuh-analysisd: ERROR: dbsync: Bad response from database: Cannot save Syscheck
> ```
> 
> The same logs appear from time to time when a syscollector event is received:
> 
> ```
> 2024/04/15 11:17:05 wazuh-modulesd:syscollector[134575] logging_helper.c:40 at taggedLogFunction(): DEBUG: Sync sent: {"component":"syscollector_packages","data":{"attributes":{"architecture":"amd64","checksum":"36ae234018ac4f6e7971195df49f6ee0d5495b24","description":"Foreign Function Interface library (development files)","format":"deb","groups":"libdevel","install_time":" ","item_id":"46336a8173721aec176faf860cedfd16a96bb917","location":" ","multiarch":"same","name":"libffi-dev","priority":"optional","scan_time":"2024/04/15 09:17:05","size":328,"source":"libffi","vendor":"Ubuntu Developers <ubuntu-devel-discuss@lists.ubuntu.com>","version":"3.4.2-4"},"index":"46336a8173721aec176faf860cedfd16a96bb917","timestamp":""},"type":"state"}
> 2024/04/15 11:17:06 wazuh-db: ERROR: DB((null)) Can't begin transaction.
> 2024/04/15 11:17:06 wazuh-db: WARNING: DB(queue/db/(null).db) not found. This behavior is unexpected, the database will be recreated.
> 2024/04/15 11:17:06 wazuh-analysisd: ERROR: dbsync: Bad response from database: Cannot save Syscheck
> 2024/04/15 11:17:06 wazuh-db: ERROR: DB((null)) Can't begin transaction.
> 2024/04/15 11:17:06 wazuh-db: WARNING: DB(queue/db/(null).db) not found. This behavior is unexpected, the database will be recreated.
> 2024/04/15 11:17:06 wazuh-analysisd: ERROR: dbsync: Bad response from database: Cannot save Syscheck
> 2024/04/15 11:17:06 wazuh-db: ERROR: DB((null)) Can't begin transaction.
> 2024/04/15 11:17:06 wazuh-db: WARNING: DB(queue/db/(null).db) not found. This behavior is unexpected, the database will be recreated.
> 2024/04/15 11:17:06 wazuh-analysisd: ERROR: dbsync: Bad response from database: Cannot save Syscheck
> 2024/04/15 11:17:06 wazuh-modulesd:syscollector[134575] logging_helper.c:40 at taggedLogFunction(): DEBUG: Sync sent: {"component":"syscollector_packages","data":{"attributes":{"architecture":"amd64","checksum":"bcae9789a25252498690b334a6a209264d623e9f","description":"multimedia player and streamer (extra access plugins)","format":"deb","groups":"video","install_time":" ","item_id":"4651e238bf5498e77c218f44b18fe0a9fa5d4b66","location":" ","multiarch":"same","name":"vlc-plugin-access-extra","priority":"optional","scan_time":"2024/04/15 09:17:05","size":223,"source":"vlc","vendor":"Ubuntu Developers <ubuntu-devel-discuss@lists.ubuntu.com>","version":"3.0.16-1build7"},"index":"4651e238bf5498e77c218f44b18fe0a9fa5d4b66","timestamp":""},"type":"state"}
> 2024/04/15 11:17:06 wazuh-modulesd:syscollector[134575] logging_helper.c:40 at taggedLogFunction(): DEBUG: Sync sent: {"component":"syscollector_packages","data":{"attributes":{"architecture":" ","checksum":"a31ef183e3dcab80d327a901e66d2571cd6b33ea","description":" ","format":"pypi","groups":" ","install_time":" ","item_id":"465dc313b5af8e21441bdaa08ee011607cf28f08","location":"/home/sebas/.local/lib/python3.10/site-packages/pyspnego-0.10.2.dist-info/METADATA","name":"pyspnego","priority":" ","scan_time":"2024/04/15 09:17:06","size":0,"source":" ","vendor":" ","version":"0.10.2"},"index":"465dc313b5af8e21441bdaa08ee011607cf28f08","timestamp":""},"type":"state"}
> 2024/04/15 11:17:06 wazuh-modulesd:syscollector[134575] logging_helper.c:40 at taggedLogFunction(): DEBUG: Sync sent: {"component":"syscollector_packages","data":{"attributes":{"architecture":" ","checksum":"4afa49081c3424f47ed395a21012b5d4927a4cdc","description":" ","format":"pypi","groups":" ","install_time":" ","item_id":"46615767238e9b3dc576ee4a827ce3322981af72","location":"/usr/lib/python3/dist-packages/webencodings-0.5.1.egg-info/PKG-INFO","name":"webencodings","priority":" ","scan_time":"2024/04/15 09:17:06","size":0,"source":" ","vendor":" ","version":"0.5.1"},"index":"46615767238e9b3dc576ee4a827ce3322981af72","timestamp":""},"type":"state"}
> 2024/04/15 11:17:06 wazuh-modulesd:syscollector[134575] logging_helper.c:40 at taggedLogFunction(): DEBUG: Sync sent: {"component":"syscollector_packages","data":{"attributes":{"architecture":"amd64","checksum":"b075e88277abd1aef92c7c67333cbf8785250399","description":"Asynchronous event notification library (core)","format":"deb","groups":"libs","install_time":" ","item_id":"467ecaf0d88545a68e338f7f2a75bd7b7acb52d8","location":" ","multiarch":"same","name":"libevent-core-2.1-7","priority":"optional","scan_time":"2024/04/15 09:17:06","size":256,"source":"libevent","vendor":"Balint Reczey <rbalint@ubuntu.com>","version":"2.1.12-stable-1build3"},"index":"467ecaf0d88545a68e338f7f2a75bd7b7acb52d8","timestamp":""},"type":"state"}
> 2024/04/15 11:17:06 wazuh-modulesd:syscollector[134575] logging_helper.c:40 at taggedLogFunction(): DEBUG: Sync sent: {"component":"syscollector_packages","data":{"attributes":{"architecture":" ","checksum":"b8696ac43bd53a3522f3263a13e99314d600510a","description":" ","format":"pypi","groups":" ","install_time":" ","item_id":"4687cb839aca0f56a725f900a61b6164f9ee4cac","location":"/usr/lib/python3/dist-packages/urllib3-1.26.5.egg-info/PKG-INFO","name":"urllib3","priority":" ","scan_time":"2024/04/15 09:17:06","size":0,"source":" ","vendor":" ","version":"1.26.5"},"index":"4687cb839aca0f56a725f900a61b6164f9ee4cac","timestamp":""},"type":"state"}
> 2024/04/15 11:17:06 wazuh-db: WARNING: DB(queue/db/(null).db) not found. This behavior is unexpected, the database will be recreated.
> 2024/04/15 11:17:06 wazuh-analysisd: ERROR: dbsync: Bad response from database: Cannot save Syscollector
> ```

Probably related to https://github.com/wazuh/wazuh/issues/22847

**Comment by sebasfalcone** [comment]
## Steps to reproduce global DB error
- Install by sources
- With the following script, download the compressed DB:
```
VD_FILENAME='vd_1.0.0_vd_4.8.0.tar.xz'
INSTALL_DIR=/var/ossec

echo "Download ${VD_FILENAME} file"
wget --no-clobber -O ${VD_FILENAME} http://packages.wazuh.com/deps/vulnerability_model_database/${VD_FILENAME}

cp ${VD_FILENAME} /var/ossec/tmp
```
- Initialize the manager

**Comment by GabrielEValenzuela** [comment]
# Analysis result

# Test Case I - Database already present 🟢

Steps

- Install from source
- Decompress the file manually with the `tar` command
- Start manager

Result: 🟢 All agents scanned, with restart of the module meanwhile.

# Test Case II - Descompress database 🔴

Steps 
- Install from source
- Download the database and copy it to `/var/ossec/tmp`
- Start manager

Result: 🔴 No errors showed during this test case, but, the scanner of the agents was random, starting with agent 000 and following with agent 006, no other agents were scanned. A lot of saved query logs were found during this test case.

**Comment by sebasfalcone** [comment]
## Test results 
- Used commit f5e08e4457418b3e243211602257cceab0da0802

### Baseline
Performed a scan from the compressed content, the same result was seen in 4.8.0
image

### Case 1 - ON / OFF Manager
We expect a cleanup on the manager
image


### Case 2 - OFF / ON Manager
We expect a re-scan on the manager
image


### Case 3 - Restart on scan agent
image

**Review by sebasfalcone — APPROVED** [review]
LGTM!
- Code 🟢 
- Tests 🟢

**Review by GabrielEValenzuela — APPROVED** [review]
Great job Octa!

**Pull request merged**